lib/Sema/DebuggerTestingTransform.cpp

```diff
@@ -0,0 +1,245 @@
+    return NextDiscriminator++;
+  }
+};
+
+/// Instrument decls with sanity-checks which the debugger can evaluate.
+class DebuggerTestingTransform : public ASTWalker {
+  ASTContext &Ctx;
+  DiscriminatorFinder &DF;
+  std::vector<DeclContext *> LocalDeclContextStack;
+
+public:
+  DebuggerTestingTransform(ASTContext &Ctx, DiscriminatorFinder &DF)
+      : Ctx(Ctx), DF(DF) {}
+
+  bool walkToDeclPre(Decl *D) override {
+    pushLocalDeclContext(D);
+
+    // Whitelist the kinds of decls to transform.
+    // TODO: Expand the set of decls visited here.
+    if (auto *FD = dyn_cast<AbstractFunctionDecl>(D))
+      return !FD->isImplicit() && FD->getBody();
+    else if (auto *TLCD = dyn_cast<TopLevelCodeDecl>(D))
+      return !TLCD->isImplicit() && TLCD->getBody();
+    return false;
+  }
+
+  bool walkToDeclPost(Decl *D) override {
+    popLocalDeclContext(D);
+    return true;
+  }
+
+  std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+    pushLocalDeclContext(E);
+
+    // Whitelist the kinds of exprs to transform.
+    // TODO: Expand the set of exprs visited here.
+    if (auto *AE = dyn_cast<AssignExpr>(E))
+      return transform(AE, AE->getDest());
+
+    return {true, E};
+  }
+
+  Expr *walkToExprPost(Expr *E) override {
+    popLocalDeclContext(E);
+    return E;
+  }
+
+private:
+  /// Return N as a local DeclContext if possible, or return nullptr.
+  DeclContext *getLocalDeclContext(ASTNode N) const {
+    DeclContext *DC = N.getAsDeclContext();
+    return (DC && DC->isLocalContext()) ? DC : nullptr;
+  }
+
+  /// If N is a local DeclContext, push it onto the context stack.
+  void pushLocalDeclContext(ASTNode N) {
+    if (auto *LDC = getLocalDeclContext(N))
+      LocalDeclContextStack.push_back(LDC);
+  }
+
+  /// If N is a local DeclContext, pop it off the context stack.
+  void popLocalDeclContext(ASTNode N) {
+    if (getLocalDeclContext(N))
+      LocalDeclContextStack.pop_back();
+  }
+
+  /// Get the current local DeclContext. This is used to create closures in the
+  /// right context.
+  DeclContext *getCurrentDeclContext() const {
+    assert(!LocalDeclContextStack.empty() && "Missing decl context");
+    return LocalDeclContextStack.back();
+  }
+
+  /// Try to extract a DeclRefExpr from the expression.
+  DeclRefExpr *extractDecl(Expr *E) {
+    while (!isa<DeclRefExpr>(E)) {
+      // TODO: Try more ways to extract interesting decl refs.
+      if (auto *Subscript = dyn_cast<SubscriptExpr>(E))
+        E = Subscript->getBase();
+      else if (auto *InOut = dyn_cast<InOutExpr>(E))
+        E = InOut->getSubExpr();
+      else
+        return nullptr;
+    }
+    return cast<DeclRefExpr>(E);
+  }
+
+  /// Attempt to create a functionally-equivalent replacement for OriginalExpr,
+  /// given that DstExpr identifies the target of some mutating update, and that
+  /// DstExpr is a subexpression of OriginalExpr.
+  std::pair<bool, Expr *> transform(Expr *OriginalExpr, Expr *DstExpr) {
```

Comment:
It wasn't immediately obvious to me that `fst` in the pair returns `true` if the transformation doesn't succeed (and false otherwise). Maybe we should add that to the comment on top of the function for clarity?

Reply:
Will fix.